### PR TITLE
Start adopting Validation codegen

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR is the first of the seria brining code-generation of validation to life.

Here's the sequence of changes that we made in relation to this:

1. The validation codegen was turned on via removing `skipValidation()` calls in build files. 
2. Several issues in codegen were discovered.
3. The issues were fixed and tested via local deployment of Validation library and subsequent update and local deployment of McJava.
5. Then, the `skipValidation()` was turned back on, and McJava version rolled back to `2.0.0-SNAPSHOT.170`, so that we can build it at GitHub. 
6. This, of course, resulted in usage of older code of Validation, which required to disable tests that were written for newer validation code.

### Further actions
 * Publish artifacts of this PR.
 * Assemble new version of McJava which uses this version of Validation.
 * Extend McJava allowing to use custom version of Validation, so that we can use McJava with the locally deployed version being built.
 * Turn test annotated with `@Disabled("Until 'skipValidation()` is turned off.")` back on.

### Other notable changes
 * Support of external validation constraints was removed.
 * Tests that required updates because of changes in codegen, were migrated to Kotlin.
 * A temporary `Workaround.kt` was introduced to allow using run-time validation for features that are not yet covered by codegen.
 * `Diags` class was introduced for gathering diagnostic messages and routines in one place. This is needed for easier introduction of localization in the future.
 * Kotlin proto DSL was adopted in the modified Kotlin code.
 * `codeBlock { }` DSL was adopted.
 * Stub proto types were separated by a file per test suite.
 * Internal and external dependencies were updated.

### CODE REVIEW

File: java-runtime/src/test/kotlin/io/spine/validate/EnclosedMessageValidationSpec.kt [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        val innerViolation = innerViolations[0]
+
+        innerViolation.msgFormat shouldStartWith Diags.Regex.prefix
+
+        assertFieldPathIs(
+            innerViolation,
+            EMAIL
+        )
+
+        innerViolation.violationList.shouldBeEmpty()
```
Comment (yevhenii-nadtochii) [inline java-runtime/src/test/kotlin/io/spine/validate/EnclosedMessageValidationSpec.kt:121]: There are many empty lines, and it is quite hard to understand grouping. Maybe reduce number of empty lines or extract smaller asserting methods?

File: java-runtime/src/test/java/io/spine/validate/option/RequiredTest.kt [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import io.spine.validate.ValidationOfConstraintTest.Companion.VALIDATION_SHOULD
+import io.spine.validate.given.MessageValidatorTestEnv
+import io.spine.validate.given.MessageValidatorTestEnv.newByteString
+import io.spine.validate.given.MessageValidatorTestEnv.newStringValue
+import org.junit.jupiter.api.Disabled
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Test
+
+@DisplayName(VALIDATION_SHOULD + "analyze `(required)` option and")
+internal class RequiredTest : ValidationOfConstraintTest() {
```
Comment (yevhenii-nadtochii) [inline java-runtime/src/test/java/io/spine/validate/option/RequiredTest.kt:52]: Should it have been called `RequiredSpec`?

File: java-runtime/src/test/java/io/spine/validate/option/RequiredTest.kt [hunk 3]
```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2022, TeamDev. All rights reserved.
```
Comment (yevhenii-nadtochii) [inline java-runtime/src/test/java/io/spine/validate/option/RequiredTest.kt:2]:
```suggestion
 * Copyright 2023, TeamDev. All rights reserved.
```

Comment (yevhenii-nadtochii) [reply java-runtime/src/test/java/io/spine/validate/option/RequiredTest.kt:2]: Though, most files have `2023` already.

File: java-runtime/src/test/java/io/spine/validate/option/RequiredTest.kt [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package io.spine.validate.option
+
```
Comment (yevhenii-nadtochii) [inline java-runtime/src/test/java/io/spine/validate/option/RequiredTest.kt:27]: We should separate `package` declaration from a copyright notice.

File: java-runtime/src/test/kotlin/io/spine/validate/NumberRangeSpec.kt [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    fun `find out that number is equal to max inclusive`() =
+        maxNumberTest(EQUAL_MAX, inclusive = true, valid = true)
+
+    @Test
+    fun `find out that number is less than max inclusive`() =
+        maxNumberTest(LESS_THAN_MAX, inclusive = true, valid = true)
+
+    @Test
+    fun `find out that number is less than max non-inclusive`() =
+        maxNumberTest(GREATER_THAN_MAX, inclusive = false, valid = false)
```
Comment (yevhenii-nadtochii) [inline java-runtime/src/test/kotlin/io/spine/validate/NumberRangeSpec.kt:97]: Can it be more straightforward?

```kotlin
maxNumberTest(LESS_THAN_MAX, inclusive = false, valid = true)
```

File: java-runtime/src/test/kotlin/io/spine/validate/option/GoesSpec.kt [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+        @Test
+        fun `are both optional` () = assertValid {
+            payment {
+                description = "Scheduled payment"
+            }
+        }
+
+        @Test
+        fun `not filled simultaneously` () {
```
Comment (yevhenii-nadtochii) [inline java-runtime/src/test/kotlin/io/spine/validate/option/GoesSpec.kt:59]:
```suggestion
        fun `not filled separately ` () {
```

File: java-runtime/src/test/java/io/spine/validate/option/RequiredTest.kt (hunk 4 above)
Comment (yevhenii-nadtochii) [reply java-runtime/src/test/java/io/spine/validate/option/RequiredTest.kt:27]: And there are many such files down.

**Review by yevhenii-nadtochii — COMMENTED** [review]
Please see my comments.

Although, I haven't viewed all files yet.

**Review by yevhenii-nadtochii — APPROVED** [review]
LGTM in general.

I've left minor suggestions to consider before merging.

File: java/src/main/kotlin/io/spine/validation/java/ValidationConstraintsCode.kt [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+    /**
+     * Obtains the generated code block.
+     */
+    fun codeBlock(): CodeBlock = code.build()
+
+    /**
+     * Obtains class-level declarations used in the validation code as code lines.
+     */
+    fun supportingMembersCode(): ImmutableList<String> {
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply java/src/main/kotlin/io/spine/validation/java/ValidationConstraintsCode.kt:90]: It's not really. The method is used only once in `ValidationCode`, so I'm going to change it to `Iterable` instead.

File: model/src/main/kotlin/io/spine/validation/NumberRules.kt [hunk 8]
```diff
@@ -97,34 +106,28 @@ private constructor(
...
+            fun Value.str(): String = when (kindCase) {
+                DOUBLE_VALUE -> doubleValue.toString()
+                INT_VALUE -> intValue.toString()
+                else -> error("Unexpected Value: `$this`.")
+            }
+
+            return "The number must be between ${lowerBound!!.str()} " +
+                    "(${lowerInclusive.str()}) and ${upperBound!!.str()} " +
+                    "(${uppedInclusive.str()}), but was {value}."
+        }
```
[the remark this answers is not in the document]
Comment (alexander-yevsyukov) [reply model/src/main/kotlin/io/spine/validation/NumberRules.kt:130]: It is a property in a sense that its value does not change. It's hardly wired to other `val` properties of this instance.

I admit it's not a trivial piece of code, but it doesn't change the nature of what's going on there. 

I'll make it `by lazy` to highlight the "propertness".

File: java-runtime/src/test/kotlin/io/spine/validate/NumberRangeSpec.kt (hunk 5 above)
Comment (alexander-yevsyukov) [reply java-runtime/src/test/kotlin/io/spine/validate/NumberRangeSpec.kt:98]: The first parameter is a value we pass to tests. The `valid` parameter when `true` tells that failure should be asserted.

Here we pass invalid value, and want an error to be raised.

**Pull request merged**